### PR TITLE
Add DictionaryMatchingOp

### PR DESCRIPTION
**Pull request by fzimmermann89**
Draft of dictionary matching function

Does only non-differentiable argmax matching, but good enough for now.
Needs tests, example, docstrings etc..

Related #465

### CODE REVIEW

**Comment by fzimmermann89** [comment]
@rkcatarina : If you join this weeks hackathon, would you be interested to work on this?

File: examples/example_DictionaryMatchOp.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+        self._f = generating_function
+        self.x: list[torch.Tensor] = []
+        self.y = torch.tensor([])
+
+    def append(self, *x: Unpack[Tin]) -> Self:
+        (newy,) = self._f(*x)
+        newy = newy.to(dtype=torch.complex64)
+        newy = newy / torch.linalg.norm(newy, dim=0, keepdim=True)
+        newy = newy.flatten(start_dim=1)
+        newx = x[1]
```
Comment (fzimmermann89) [inline examples/example_DictionaryMatchOp.py]: Save all the x values, not only the second parameter

So self.x is a list of all the parameters:
Here: [M0,T1]

File: examples/example_DictionaryMatchOp.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+class DictionaryMatchOp(Operator[torch.Tensor, tuple[*Tin]]):
+    def __init__(self, generating_function: Callable[[Unpack[Tin]], tuple[torch.Tensor,]]):
+        super().__init__()
+        self._f = generating_function
+        self.x: list[torch.Tensor] = []
+        self.y = torch.tensor([])
+
+    def append(self, *x: Unpack[Tin]) -> Self:
+        (newy,) = self._f(*x)
+        newy = newy.to(dtype=torch.complex64)
```
Comment (fzimmermann89) [inline examples/example_DictionaryMatchOp.py]: Keep the dtype of y

File: examples/example_DictionaryMatchOp.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+            self.y = newy
+            return self
+        self.x = [torch.cat(old, new) for old, new in zip(self.x, newx, strict=True)]
+        self.y = torch.cat((self.y, newy))
+        return
+
+    def forward(self, input_signal: torch.Tensor) -> tuple[Unpack[Tin]]:
+        similar = einops.einsum(input_signal, self.y, 't ..., t idx -> idx ...')
+        idx = torch.argmax(torch.abs(similar), dim=0)
+        t1_start_neu = self.x[idx]
```
Comment (fzimmermann89) [inline examples/example_DictionaryMatchOp.py]: Return instead more than one parameter
x=tuple(xi[idx] for xi in self.x)
Return x

**Comment by fzimmermann89** [comment]
This already looks promising :)

Why did you change from the list of x values (in the initial commit here) to only supporting a single T1?

It would be great if the matching op can be used with multiple parameters (fingerprinting for example)

File: src/mrpro/operators/DictionaryMatchOp.py [hunk 4]
```diff
@@ -53,7 +53,11 @@ def append(self, *x: Unpack[Tin]) -> Self:
 
         """
         (newy,) = self._f(*x)
+        newy = newy.to(dtype=torch.complex64)
```
Comment (rkcatarina) [inline src/mrpro/operators/DictionaryMatchOp.py]: The signal_dictionary has to be complex before being used in line 90 in "einops.einsum". Making it complex inside the "einops.einsum" results in wrong results.

File: src/mrpro/operators/DictionaryMatchOp.py [hunk 5]
```diff
@@ -53,7 +53,11 @@ def append(self, *x: Unpack[Tin]) -> Self:
...
         newy = newy / torch.linalg.norm(newy, dim=0, keepdim=True)
+
+        if newy.ndimension() == 1:
```
Comment (rkcatarina) [inline src/mrpro/operators/DictionaryMatchOp.py]: Deleted in next commit.

File: src/mrpro/operators/DictionaryMatchOp.py [hunk 6]
```diff
@@ -83,7 +87,7 @@ def forward(self, input_signal: torch.Tensor) -> tuple[Unpack[Tin]]:
         """
         if not self.x:
             raise KeyError('No keys in the dictionary. Please first add some x values using `append`.')
-        similarity = einops.einsum(input_signal, self.y.conj(), 'm ..., m idx  -> idx ...').abs()
+        similarity = einops.einsum(input_signal, self.y, 'm ..., m idx  -> idx ...').abs()
```
Comment (rkcatarina) [inline src/mrpro/operators/DictionaryMatchOp.py]: see comment for line 56

File: src/mrpro/operators/DictionaryMatchOp.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    def __init__(
+        self, generating_function: Callable[[Unpack[Tin]], tuple[torch.Tensor,]], scaling_argument: int | None = None
+    ):
+        """Initialize DictionaryMatchOp.
+
+        Parameters
+        ----------
+        generating_function
+            signal model that takes n inputs and returns a signal y.
+        scaling_argument
```
Comment (ckolbPTB) [inline src/mrpro/operators/DictionaryMatchOp.py]: index_of_scaling_parameter
Dictionary matching is insensitive to overall signal scaling as all the signal curves are normalized. A scaling factor (e.g. the equilibrium magnetisation`m0` in `~mrpro.operators.models.InversionRecovery`) is calculated after the dictionary matching if `index_of_scaling_parameter` is not `None`. `index_of_scaling_parameter` then needs to be set to the index of the scaling parameter in the signal model. For `~mrpro.operators.models.InversionRecovery` the parameters are `[m0, t1]` and therefore `index_of_scaling_parameter` should be set to ``0``. The operator will then return `t1` estimated via dictionary matching and `m0` via a post-processing step. If `index_of_scaling_parameter` is `None`, `m0` will be ``1``.

Comment (ckolbPTB) [reply src/mrpro/operators/DictionaryMatchOp.py]: A bit long but hopefully understandable..

Comment (fzimmermann89) [reply src/mrpro/operators/DictionaryMatchOp.py]: Thanks! Still not perfect, but way better!

File: src/mrpro/operators/DictionaryMatchOp.py (hunk 4 above)
Comment (fzimmermann89) [reply src/mrpro/operators/DictionaryMatchOp.py]: do you have some example code where both ways are compared? 

at least the example currently seems to produce good results

**Comment by fzimmermann89** [comment]
@rkcatarina 
I f'ed up git a bit. So this was force pushed.
When pulling in the changes, make sure to either use `git pull --rebase --autostash`  if you have useful changes locally.
If you don't have useful local changes in the branch, do
```
git fetch
git reset --hard origin/dict
```
to reset your local state to what is on github

File: tests/operators/test_dictionary_matching_op.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    operator.append(m0, t1)
+
+    m0_matched, t1_matched = operator(y)
+
+    if scaling_argument is not None:
+        torch.testing.assert_close(m0_matched, m0, atol=1e-3, rtol=0.0)
+    torch.testing.assert_close(t1_matched, t1, atol=1e-4, rtol=0.0)
+
+
+# TODO:
```
Comment (fzimmermann89) [inline tests/operators/test_dictionary_matching_op.py]: Some todos

File: tests/operators/test_dictionary_matching_op.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+@pytest.mark.parametrize('shape', [(5, 4, 3)])
+@pytest.mark.parametrize('index_of_scaling_parameter', [None, 0], ids=['dont_predict_scale', 'predict_scale'])
+
+def test_dictionaryop_append(shape: tuple[int], dtype: torch.dtype, index_of_scaling_parameter: int | None) -> None:
+    """Test dictionary matching with concatenated entries."""
+    rng_1 = RandomGenerator(2)
+    m0_1 = rng_1.rand_tensor(shape, dtype=dtype, low=0.2, high=1.0)
+    t1_1 = rng_1.rand_tensor(shape, dtype=dtype.to_real(), low=0.1, high=1.0)
+
+    rng_2 = RandomGenerator(3)
```
Comment (fzimmermann89) [inline tests/operators/test_dictionary_matching_op.py]: you can reuse the same rng

File: tests/operators/test_dictionary_matching_op.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    m0_2 = rng_2.rand_tensor(shape, dtype=dtype, low=0.2, high=1.0)
+    t1_2 = rng_2.rand_tensor(shape, dtype=dtype.to_real(), low=0.1, high=1.0)
+
+    #concatenation of the tensors
+    m0_cat = torch.cat((m0_1, m0_2))
```
Comment (fzimmermann89) [inline tests/operators/test_dictionary_matching_op.py:57]: move this to the     #dictionary matching when appending the individual tensors section

File: tests/operators/test_dictionary_matching_op.py [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+
+    rng = RandomGenerator(2)
+    model = InversionRecovery(rng.float32_tensor(5))
+    m0 = rng.rand_tensor(shape, dtype=dtype, low=0.2, high=1.0)
+    t1 = rng.rand_tensor(shape, dtype=dtype.to_real(), low=0.1, high=1.0)
+    (y,) = model(m0, t1)
+    operator = DictionaryMatchOp(model, index_of_scaling_parameter=index_of_scaling_parameter)
+    operator.append(m0, t1)
+
+    if index_of_scaling_parameter is not None:
```
Comment (fzimmermann89) [inline tests/operators/test_dictionary_matching_op.py]: just remove this test.

Comment (fzimmermann89) [reply tests/operators/test_dictionary_matching_op.py]: the cutrrent version does not actually test behavior of the operator.

and in the operator we actually currently seem to support any value for the index, and just wrap it around. 
so even for a two parameter model, you could set it to 1000.
there is no really nice way to check it when creating the operator, only later on...

File: tests/operators/test_dictionary_matching_op.py [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+
+import pytest
+import torch
+from mrpro.operators import DictionaryMatchOp
+from mrpro.operators.models import InversionRecovery
+
+from tests import RandomGenerator
+
+
+@pytest.mark.parametrize('dtype', [torch.float32, torch.complex64], ids=['float32', 'complex64'])
```
Comment (fzimmermann89) [inline tests/operators/test_dictionary_matching_op.py:11]: add here a "-2" case for  index_of_scaling_parameter , name it predict_scale_negative_index

**Review by fzimmermann89 — COMMENTED** [review]
some comments

File: examples/scripts/qmri_sg_challenge_2024_t1.py [hunk 13]
```diff
@@ -222,8 +209,9 @@ def show_images(*images: torch.Tensor, titles: list[str] | None = None) -> None:
 # as $q(TE) = M_0 e^{-TE/T_2^*}$ with the equilibrium magnetization $M_0$, the echo time $TE$, and $T_2^*$.\
 # Give it a try and see if you can obtain good $T_2^*$ maps!
 # ```{note}
-# The echo times $TE$ can be found in `IData.header.te`. A good starting value for $M_0$ is the signal at the shortest
-# echo time. A good starting value for $T_2^*$ is 20 ms.
+# The echo times $TE$ can be found in `IData.header.te`. Instead of using dictionary matching for starting values,
```
Comment (ckolbPTB) [inline examples/scripts/qmri_sg_challenge_2024_t1.py]: This sounds as if dictionary matching would not be a good starting value. Maybe write something like "Either dictionary matching can be used to get starting values or ... "

File: src/mrpro/operators/DictionaryMatchOp.py [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+        self.inverse_norm_y = None if index_of_scaling_parameter is None else torch.tensor([])
+
+    def append(self, *x: Unpack[Tin]) -> Self:
+        """Append `x` values to the dictionary.
+
+        Parameters
+        ----------
+        x
+            points where the signal model will be evaluated. For signal models
+            with n inputs, n Tensors should be provided. Broadcasting is supported.
```
Comment (ckolbPTB) [inline src/mrpro/operators/DictionaryMatchOp.py]:
```suggestion
            with n inputs, n tensors should be provided. Broadcasting is supported.
```

File: src/mrpro/operators/DictionaryMatchOp.py [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+
+        Returns
+        -------
+        Self
```
Comment (ckolbPTB) [inline src/mrpro/operators/DictionaryMatchOp.py]: dictionary? Whatever the term, it should be indented

File: tests/operators/test_dictionary_matching_op.py [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+@pytest.mark.parametrize('shape', [(5, 4, 3)])
+@pytest.mark.parametrize(
+    'index_of_scaling_parameter',
+    [None, -2, 0],
+    ids=['dont_predict_scale', 'predict_scale_negative_index', 'predict_scale'],
+)
+def test_dictionaryop_matching(shape: tuple[int], dtype: torch.dtype, index_of_scaling_parameter: int | None) -> None:
+    """Test dictionary matching with real-valued signals."""
```
Comment (ckolbPTB) [inline tests/operators/test_dictionary_matching_op.py]: why real-valued? isn't the input also complex?

File: tests/operators/test_dictionary_matching_op.py [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+    m0_1 = rng.rand_tensor(shape, dtype=dtype, low=0.2, high=1.0)
+    t1_1 = rng.rand_tensor(shape, dtype=dtype.to_real(), low=0.1, high=1.0)
+
+    m0_2 = rng.rand_tensor(shape, dtype=dtype, low=0.2, high=1.0)
+    t1_2 = rng.rand_tensor(shape, dtype=dtype.to_real(), low=0.1, high=1.0)
+
+    model = InversionRecovery(rng.float32_tensor(5))
+
+    # dictionary matching when appending the individual tensors
+    # concatenation of the tensors
```
Comment (ckolbPTB) [inline tests/operators/test_dictionary_matching_op.py]:
```suggestion
```

File: tests/operators/test_dictionary_matching_op.py [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+    [None, -2, 0],
+    ids=['dont_predict_scale', 'predict_scale_negative_index', 'predict_scale'],
+)
+def test_dictionaryop_no_entry(index_of_scaling_parameter: int | None) -> None:
+    """Test dictionary matching when no entries have been appended before."""
+    rng = RandomGenerator(2)
+    model = InversionRecovery(rng.float32_tensor(5))
+
+    operator = DictionaryMatchOp(model, index_of_scaling_parameter=index_of_scaling_parameter)
+    # create empty signal model
```
Comment (ckolbPTB) [inline tests/operators/test_dictionary_matching_op.py]:
```suggestion
```

File: tests/operators/test_dictionary_matching_op.py [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+
+
+@pytest.mark.parametrize('dtype', [torch.float32, torch.complex64], ids=['float32', 'complex64'])
+@pytest.mark.parametrize('shape', [(5, 4, 3)])
+@pytest.mark.parametrize(
+    'index_of_scaling_parameter',
+    [None, -2, 0],
+    ids=['dont_predict_scale', 'predict_scale_negative_index', 'predict_scale'],
+)
+def test_dictionaryop_matching(shape: tuple[int], dtype: torch.dtype, index_of_scaling_parameter: int | None) -> None:
```
Comment (ckolbPTB) [inline tests/operators/test_dictionary_matching_op.py]:
```suggestion
def test_dictionary_matching_op(shape: tuple[int], dtype: torch.dtype, index_of_scaling_parameter: int | None) -> None:
```

File: tests/operators/test_dictionary_matching_op.py [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+
+
+@pytest.mark.parametrize('dtype', [torch.float32, torch.complex64], ids=['float32', 'complex64'])
+@pytest.mark.parametrize('shape', [(5, 4, 3)])
+@pytest.mark.parametrize(
+    'index_of_scaling_parameter',
+    [None, -2, 0],
+    ids=['dont_predict_scale', 'predict_scale_negative_index', 'predict_scale'],
+)
+def test_dictionaryop_append(shape: tuple[int], dtype: torch.dtype, index_of_scaling_parameter: int | None) -> None:
```
Comment (ckolbPTB) [inline tests/operators/test_dictionary_matching_op.py]:
```suggestion
def test_dictionary_matching_op_append(shape: tuple[int], dtype: torch.dtype, index_of_scaling_parameter: int | None) -> None:
```

File: tests/operators/test_dictionary_matching_op.py [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+        torch.testing.assert_close(m0_matched, m0_matched_cat, atol=1e-3, rtol=0.0)
+    torch.testing.assert_close(t1_matched, t1_matched_cat, atol=1e-4, rtol=0.0)
+
+
+@pytest.mark.parametrize(
+    'index_of_scaling_parameter',
+    [None, -2, 0],
+    ids=['dont_predict_scale', 'predict_scale_negative_index', 'predict_scale'],
+)
+def test_dictionaryop_no_entry(index_of_scaling_parameter: int | None) -> None:
```
Comment (ckolbPTB) [inline tests/operators/test_dictionary_matching_op.py]:
```suggestion
def test_dictionary_matching_op_empty_dictionary(index_of_scaling_parameter: int | None) -> None:
```

**Review by ckolbPTB — APPROVED** [review]
Just a few small things.

File: src/mrpro/operators/DictionaryMatchOp.py (hunk 15 above)
Comment (fzimmermann89) [inline src/mrpro/operators/DictionaryMatchOp.py]:
```suggestion
            Self
```

File: tests/operators/test_dictionary_matching_op.py (hunk 16 above)
Comment (fzimmermann89) [inline tests/operators/test_dictionary_matching_op.py]:
```suggestion
    """Test dictionary matching."""
```

File: examples/scripts/qmri_sg_challenge_2024_t1.py [hunk 22]
```diff
@@ -225,8 +213,9 @@ def show_images(*images: torch.Tensor, titles: list[str] | None = None) -> None:
...
+# a good starting value for $M_0$ is the signal at the shortest echo time
+# and a good starting value for $T_2^*$ is 20 ms.
```
Comment (fzimmermann89) [inline examples/scripts/qmri_sg_challenge_2024_t1.py]:
```suggestion
# The echo times $TE$ can be found in `IData.header.te`. As starting values, either dictionary matching, or
# the signal at the shortest echo time for $M_0$ and 20 ms for $T_2^*$ can be used.
```
